File: projects/no1kmedi/src/lib/logosInquiryReportV1.ts
/** Logos inquiry report v1 — Standard tier S1-S5 (mirrors scripts/core/logos_inquiry_report_v1.py) */
import { createHash } from "node:crypto";

import type { StudioQueryPayload } from "./logosResearchStudioV1";
import type { ConflictContextResult } from "./logosStudioConflictBridgeV1";
import {
  evaluateLogosTextMvpIntake,
  type LogosTextMvpIntakeSummary,
} from "./logosResearchTextMvpV1";
import type { SasangRegimeHintBTrack } from "./logosInquiryFreezeMetaV1";
import { filterStudioBoilerplateLines, isStudioBoilerplateKo } from "./logosStudioBoilerplateV1";
import {
  firstNonStubLine,
  isPublicS4StubPhrase,
  PUBLIC_S4_STUB_PHRASE_RE,
} from "./logosInquiryPublicS4FilterV1";

export type LogosInquiryReportV1 = {
  schema: "logos_inquiry_report_v1";
  version: string;
  tier: "standard";
  output_format: "inquiry_report_v1";
  generated_at_utc: string;
  research_only: boolean;
  non_gating: boolean;
  send_gate: string;
  query: string;
  preset_id?: string | null;
  query_mode?: string | null;
  intake: LogosTextMvpIntakeSummary;
  sections: {
    S1: {
      section_id: "S1_citation_lock";
      title_ko: string;
      verse_refs: string[];
      citation_lock_anchors: string[];
      security_note_ko: string;
    };
    S2: {
      section_id: "S2_lexicon_hash";
      title_ko: string;
      lemma_edge_line_count: number | null;
      min_line_count_floor: number;
      floor_pass: boolean;
      freeze_manifest_pointer: string;
      manifest_sha256: string | null;
      lemma_edges_sha256: string;
      sidecar_corpus_sha256: string;
      path_token_preview: string[];
      security_note_ko: string;
    };
    S3: {
      section_id: "S3_context_divergence";
      title_ko: string;
      groups: Array<Record<string, unknown>>;
      note_ko: string;
      regime_hint_b_track?: SasangRegimeHintBTrack | null;
    };
    S4: {
      section_id: "S4_dynamic_synthesis";
      title_ko: string;
      body_ko: string;
      bullets_ko: string[];
      streaming_deferred?: "P0-1b" | "active";
      format_gate?: {
        applied: boolean;
        missing_sections: string[];
        recomposed: boolean;
      };
    };
    S5: {
      section_id: "S5_jema_integrity_signoff";
      title_ko: string;
      signoff_status?: "provisional" | "final";
      chain_exit_code: number;
      artifact_path: string;
      sections_payload_sha256: string | null;
      freeze_verify_command: string;
      signed_at_utc: string;
      note_ko?: string;
    };
  };
  governance: {
    disclaimer_ko: string;
    forbidden_claims: string[];
    quality_basis_ko: string;
  };
  evidence_confidence?: StudioQueryPayload["evidence_confidence"] | null;
  azure_distill_meta?: StudioQueryPayload["azure_distill_meta"] | null;
};

export type FreezeLexiconMeta = {
  lemma_edge_line_count: number | null;
  min_line_count_floor: number;
  floor_pass: boolean;
  freeze_manifest_pointer: string;
  manifest_sha256: string | null;
  lemma_edges_sha256: string;
  sidecar_corpus_sha256: string;
};

export const DEFAULT_FREEZE_LEXICON_META: FreezeLexiconMeta = {
  lemma_edge_line_count: 298984,
  min_line_count_floor: 290000,
  floor_pass: true,
  freeze_manifest_pointer: "docs/final/artifacts/logos_corpus_knowledge_freeze_manifest_v1_latest.json",
  manifest_sha256: null,
  lemma_edges_sha256: "b0f33b703e88d986a75baa446571242d4f097567b4eadd68b061d857685446c0",
  sidecar_corpus_sha256: "e9cdd90c278b37a0316235438a9dfe187543c66b966ac14c6c5ac19c23178c32",
};

const FREEZE_ARTIFACT = "reports/logos_inquiry_report_schema_freeze_v1_latest.json";

function stripHypo(text: string): string {
  return (text || "").replace(/^\[HYPO\]\s*/i, "").replace(/\[HYPO\]\s*/g, "").trim();
}

const GEMATRIA_BULLET_RE =
  /combined_sum|vector_4d|hub_score|state16|Gematria_Pin|topology pin|mispar_/i;

const PUBLIC_LEAK_RE =
  /lemma:gnosis:|shared_lemma=|Lemma\s*연결\s*이웃\s*구절|Path\s*envelope|orphan\s*veto|Gematria_Pin/i;

export { isPublicS4StubPhrase, PUBLIC_S4_STUB_PHRASE_RE } from "./logosInquiryPublicS4FilterV1";

function firstNonStub(lines: string[]): string {
  return firstNonStubLine(lines);
}

function prioritizeVerseRefsForQuery(query: string, verseRefs: string[]): string[] {
  const q = (query || "").trim();
  const refs = [...(verseRefs || [])].map((r) => String(r).trim()).filter(Boolean);
  if (!q) return refs;

  const isPsalm23Query =
    /시편\s*23|psalm\s*23|ps\.?\s*23/i.test(q) ||
    (/(시편|psalm)/i.test(q) && /목자|shepherd/i.test(q));

  if (isPsalm23Query) {
    const ps23 = refs.filter((r) => /^Ps\.23/i.test(r.replace(/\s/g, "")));
    const rest = refs.filter((r) => !/^Ps\.23/i.test(r.replace(/\s/g, "")));
    if (ps23.length) return [...ps23, ...rest];
    const withoutJobStub = refs.filter((r) => !/^Job\./i.test(r.replace(/\s/g, "")));
    return ["Ps.23.1", "Ps.23.4", ...withoutJobStub];
  }

  return refs;
}

function coreClaimFallback(query: string, verseRefs: string[]): string {
  const refs = prioritizeVerseRefsForQuery(query, verseRefs);
  const ps23 = refs.filter((r) => /^Ps\.23/i.test(r.replace(/\s/g, "")));
  if (ps23.length) {
    return `시편 23편(${ps23.slice(0, 2).join(" · ")})의 목자·신뢰·길 안내 이미지를 citation lock 앵커 중심으로 해석합니다.`;
  }
  if (refs.length) {
    return `${refs.slice(0, 2).join(" · ")} 앵커 구절을 중심으로 질문의 논지를 정리합니다.`;
  }
  return "질문의 중심 논지를 citation lock 앵커 기준으로 해석합니다.";
}

function scrubStubFromFiveSectionBody(body: string, query: string, verseRefs: string[]): string {
  const raw = normalizeWhitespace(body || "");
  if (!raw) return raw;

  const sectionRe = /(###\s*핵심\s*주장\s*\n)([\s\S]*?)(?=\n###\s|$)/i;
  const match = sectionRe.exec(raw);
  if (!match) return raw;

  const heading = match[1];
  const coreBody = match[2].trim();
  const coreLines = coreBody.split("\n").map((l) => l.trim()).filter(Boolean);
  const cleanCore = firstNonStub(coreLines);
  if (cleanCore && !isPublicS4StubPhrase(cleanCore)) {
    return raw.replace(sectionRe, `${heading}${cleanCore}\n`);
  }

  const fallback = coreClaimFallback(query, verseRefs);
  return raw.replace(sectionRe, `${heading}${fallback}\n`);
}

function leadAnswerSection(text: string): string {
  const raw = (text || "").trim();
  return (raw.split(/(?:^|\n|\s)---(?:\s|\n)|###\s*Reading pack/i)[0] ?? raw).trim();
}

function splitBullets(text: string): string[] {
  const bullets: string[] = [];
  for (const line of leadAnswerSection(text).split("\n")) {
    const m = line.trim().match(/^[-*•]\s+(.+)$/);
    if (m) {
      const item = stripHypo(m[1]);
      if (!GEMATRIA_BULLET_RE.test(item)) bullets.push(item);
    }
  }
  return bullets;
}

function normalizeWhitespace(text: string): string {
  return (text || "")
    .replace(/[ \t]{2,}/g, " ")
    .replace(/\n{3,}/g, "\n\n")
    .trim();
}

/** Public S4 body — strip internal research tags (display/scoring parity with Python lib). */
export function stripPublicResearchTagsForS4(body: string): string {
  return normalizeWhitespace(
    (body || "")
      .replace(/\[HYPO\]\s*/gi, "")
      .replace(/\[NON_GATING\]\s*/gi, "")
      .replace(/\bresearch_only\b/gi, "")
      .replace(/\bsend_gate\s*:\s*\w+/gi, ""),
  );
}

function stripPublicLeakSegments(text: string): string {
  return normalizeWhitespace(
    (text || "")
      .replace(/(?:^|\n)\s*Lemma\s*연결\s*이웃\s*구절[\s\S]*?(?=(?:\n###\s*Reading pack)|$)/gi, "\n")
      .replace(/lemma:gnosis:[^\s,;]+/gi, "")
      .replace(/shared_lemma=\d+/gi, "")
  );
}

export function sanitizeInquiryS4ForPublic(body: string): {
  body: string;
  leak_detected: boolean;
  fallback_applied: boolean;
} {
  const raw = String(body || "").trim();
  const leak_detected = PUBLIC_LEAK_RE.test(raw);
  if (!raw) {
    return {
      body: "Citation lock 기반 해석 요약을 준비 중입니다. 구절 앵커 중심으로 간단히 다시 질문해 주세요. [NON_GATING]",
      leak_detected: false,
      fallback_applied: true,
    };
  }
  const cleaned = stripPublicLeakSegments(raw);
  if (!leak_detected) {
    return { body: cleaned || raw, leak_detected: false, fallback_applied: false };
  }
  if (cleaned.length >= 120) {
    return { body: cleaned, leak_detected: true, fallback_applied: false };
  }
  return {
    body:
      "Citation lock 앵커 구절 기준의 해석 요약만 공개합니다. 내부 lemma raw/토큰 덤프는 숨김 처리되었습니다. 핵심 구절(예: Dan.2.10-19)의 의미 축을 지정해 다시 질문해 주세요. [NON_GATING]",
    leak_detected: true,
    fallback_applied: true,
  };
}

const FIVE_SECTION_TITLES = [
  "### 핵심 주장",
  "### 근거 구절",
  "### 반증·대안",
  "### 한계·주의",
  "### 다음 행동 제안",
] as const;

function splitSentences(text: string): string[] {
  return (text || "")
    .replace(/\s+/g, " ")
    .split(/(?<=[.!?。！？])\s+/)
    .map((s) => s.trim())
    .filter(Boolean);
}


function extractRefsFromBody(body: string): string[] {
  const refs = new Set<string>();
  const refRe = /\b(?:[1-3]\s*)?[A-Za-z가-힣]+\.?\s*\d{1,3}(?::\d{1,3}(?:-\d{1,3})?)?/g;
  for (const m of body.matchAll(refRe)) {
    const ref = String(m[0] || "").replace(/\s+/g, " ").trim();
    if (ref) refs.add(ref);
  }
  return Array.from(refs);
}

function buildFiveSectionS4(input: {
  body: string;
  bullets: string[];
  verseRefs: string[];
  query?: string;
}): {
  body: string;
  missing_sections: string[];
  recomposed: boolean;
} {
  const query = input.query ?? "";
  const body = normalizeWhitespace(input.body || "");
  const bullets = (input.bullets || [])
    .map((b) => b.trim())
    .filter((b) => b && !isPublicS4StubPhrase(b));
  const verseRefs = prioritizeVerseRefsForQuery(query, input.verseRefs ?? []);
  const hasAllSections = FIVE_SECTION_TITLES.every((title) => body.includes(title));
  if (hasAllSections) {
    return {
      body: scrubStubFromFiveSectionBody(body, query, verseRefs),
      missing_sections: [],
      recomposed: false,
    };
  }

  const lines = body.split("\n").map((line) => line.trim()).filter(Boolean);
  const sentences = splitSentences(body).filter((s) => !isPublicS4StubPhrase(s));
  const refs = (verseRefs.length ? verseRefs : extractRefsFromBody(body)).slice(0, 4);
  const coreClaim =
    firstNonStub([
      ...bullets,
      ...sentences,
      ...lines,
    ]) || coreClaimFallback(query, refs);
  const evidenceLine =
    refs.length > 0
      ? refs.map((ref) => `- ${ref}`).join("\n")
      : "- 명시 구절이 부족해 재질의가 필요합니다. (권·장·절 앵커 권장)";
  const counterLine =
    firstNonStub(bullets.slice(1).concat(sentences.slice(1))) ||
    "대안 해석 가능성을 병기하며 단일 해석을 절대화하지 않습니다.";
  const limitLine =
    firstNonStub(bullets.slice(2).concat(sentences.slice(2))) ||
    "본 응답은 연구 요약이며 교리 확정/실행 지시가 아닙니다.";
  const actionLine =
    firstNonStub(bullets.slice(3)) ||
    "핵심 구절 1~2개를 지정해 재질문하면 문맥-반증 비교를 더 정밀화할 수 있습니다.";
  const recomposed = normalizeWhitespace(
    [
      "### 핵심 주장",
      coreClaim,
      "",
      "### 근거 구절",
      evidenceLine,
      "",
      "### 반증·대안",
      counterLine,
      "",
      "### 한계·주의",
      limitLine,
      "",
      "### 다음 행동 제안",
      actionLine,
    ].join("\n"),
  );
  return {
    body: recomposed,
    missing_sections: FIVE_SECTION_TITLES.filter((title) => !body.includes(title)),
    recomposed: true,
  };
}

export function applyInquiryS4QualityGate(input: {
  body: string;
  bullets: string[];
  verseRefs: string[];
  query?: string;
}): {
  body: string;
  leak_detected: boolean;
  fallback_applied: boolean;
  missing_sections: string[];
  recomposed: boolean;
} {
  const query = input.query ?? "";
  const verseRefs = prioritizeVerseRefsForQuery(query, input.verseRefs ?? []);
  const sanitized = sanitizeInquiryS4ForPublic(input.body);
  const composed = buildFiveSectionS4({
    body: sanitized.body,
    bullets: (input.bullets ?? []).filter((b) => !isPublicS4StubPhrase(b)),
    verseRefs,
    query,
  });
  return {
    body: stripPublicResearchTagsForS4(
      scrubStubFromFiveSectionBody(composed.body, query, verseRefs),
    ),
    leak_detected: sanitized.leak_detected,
    fallback_applied: sanitized.fallback_applied,
    missing_sections: composed.missing_sections,
    recomposed: composed.recomposed,
  };
}

function ensureFiveSectionS4Draft(input: {
  body: string;
  bullets: string[];
  verseRefs: string[];
  query?: string;
}): string {
  const built = buildFiveSectionS4(input);
  return built.body;
}

function extractAnchors(payload: StudioQueryPayload): string[] {
  const anchors: string[] = [];
  const seen = new Set<string>();
  const conflict = payload.conflict_context as Extract<ConflictContextResult, { ok: true }> | null;
  if (conflict?.groups) {
    for (const group of conflict.groups) {
      for (const school of group.schools ?? []) {
        for (const anchor of school.citation_lock_anchors ?? []) {
          const t = String(anchor).trim();
          if (t && !seen.has(t)) {
            seen.add(t);
            anchors.push(t);
          }
        }
      }
    }
  }
  for (const ref of payload.path?.verse_refs ?? []) {
    const t = String(ref).trim();
    if (t && !seen.has(t)) {
      seen.add(t);
      anchors.push(t);
    }
  }
  return anchors;
}

function buildSchoolGroups(payload: StudioQueryPayload): Array<Record<string, unknown>> {
  const conflict = payload.conflict_context as Extract<ConflictContextResult, { ok: true }> | null;
  if (!conflict?.groups?.length) return [];
  return conflict.groups.map((group) => ({
    conflict_group_id: group.conflict_group_id,
    lexicon_base: group.lexicon_base,
    school_count: group.school_count ?? group.schools?.length ?? 0,
    schools: (group.schools ?? []).map((school) => ({
      school_tier: school.school_tier,
      interpretation_ko: stripHypo(school.interpretation_ko ?? ""),
      verse_refs: school.verse_refs ?? [],
      citation_lock_anchors: school.citation_lock_anchors ?? [],
      traditions: school.traditions ?? [],
    })),
  }));
}

function pathTokenPreview(payload: StudioQueryPayload, cap = 16): string[] {
  const tokens: string[] = [];
  const seen = new Set<string>();
  const push = (t: string) => {
    const v = t.trim();
    if (!v || seen.has(v)) return;
    seen.add(v);
    tokens.push(v);
  };

  for (const raw of [...(payload.path?.node_ids ?? []), ...(payload.path?.steps ?? [])]) {
    const t = String(raw).trim();
    if (/^[A-Za-z0-9]+\.\d/.test(t)) push(t);
    else if (/^lemma_/.test(t) || t.startsWith("node:")) push(t.replace(/^node:/, ""));
    else if (/gematria/i.test(t)) push(t);
    if (tokens.length >= cap) break;
  }

  for (const ref of payload.path?.verse_refs ?? []) {
    if (tokens.length >= cap) break;
    const verse = String(ref).trim();
    if (!verse) continue;
    const pin = createHash("sha256").update(`gematria_pin:${verse}`, "utf8").digest("hex").slice(0, 12);
    push(`Gematria_Pin:${pin}`);
  }

  return tokens.slice(0, cap);
}

function sortKeysDeep(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    return Object.keys(obj)
      .sort()
      .reduce<Record<string, unknown>>((acc, key) => {
        acc[key] = sortKeysDeep(obj[key]);
        return acc;
      }, {});
  }
  return value;
}

function sha256Sections(s1s4: Record<string, unknown>): string {
  const canonical = JSON.stringify(sortKeysDeep(s1s4));
  return createHash("sha256").update(canonical, "utf8").digest("hex");
}

export function buildLogosInquiryReport(
  payload: StudioQueryPayload,
  query: string,
  intake: LogosTextMvpIntakeSummary,
  freezeMeta: FreezeLexiconMeta = DEFAULT_FREEZE_LEXICON_META,
  chainExitCode = 0,
  sasangHint: SasangRegimeHintBTrack | null = null,
): LogosInquiryReportV1 {
  const answer = stripHypo(payload.answer ?? "");
  const verseRefs = (payload.path?.verse_refs ?? []).map((v) => String(v).trim()).filter(Boolean);
  const bullets = filterStudioBoilerplateLines(splitBullets(payload.answer ?? ""));
  if (payload.insight_card?.gap_ko) {
    const gap = stripHypo(payload.insight_card.gap_ko);
    if (!isStudioBoilerplateKo(gap)) bullets.push(gap);
  }

  const s1 = {
    section_id: "S1_citation_lock" as const,
    title_ko: "Citation Lock (구절 인용)",
    verse_refs: verseRefs.slice(0, 24),
    citation_lock_anchors: extractAnchors(payload).slice(0, 24),
    security_note_ko: "canonical verse ref·citation lock 인덱스만 — KRV/31k 원문 전문 노출 금지.",
  };
  const s2 = {
    section_id: "S2_lexicon_hash" as const,
    title_ko: "Lexicon Hash (원어 렉시콘)",
    lemma_edge_line_count: freezeMeta.lemma_edge_line_count,
    min_line_count_floor: freezeMeta.min_line_count_floor,
    floor_pass: freezeMeta.floor_pass,
    freeze_manifest_pointer: freezeMeta.freeze_manifest_pointer,
    manifest_sha256: freezeMeta.manifest_sha256,
    lemma_edges_sha256: freezeMeta.lemma_edges_sha256,
    sidecar_corpus_sha256: freezeMeta.sidecar_corpus_sha256,
    path_token_preview: pathTokenPreview(payload),
    security_note_ko: "29만 행 원문 덤프 금지 — line_count·sha256 pin·경로 토큰 preview만.",
  };
  const s3 = {
    section_id: "S3_context_divergence" as const,
    title_ko: "Context Divergence (맥락/학파 분기)",
    groups: buildSchoolGroups(payload),
    note_ko: "[NON_GATING] conflict_context 기반 — Track A·실매매 트리거 아님.",
    ...(sasangHint ? { regime_hint_b_track: sasangHint } : {}),
  };
  const s4 = {
    section_id: "S4_dynamic_synthesis" as const,
    title_ko: "Dynamic Synthesis (고차원 통찰)",
    body_ko: ensureFiveSectionS4Draft({
      body: answer || "응답 본문을 생성하지 못했습니다.",
      bullets,
      verseRefs,
      query: query.trim(),
    }),
    bullets_ko: bullets.slice(0, 8),
    streaming_deferred: "active" as const,
  };
  const s1s4 = { S1: s1, S2: s2, S3: s3, S4: s4 };
  const signedAt = new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
  const s5 = {
    section_id: "S5_jema_integrity_signoff" as const,
    title_ko: "JEMA Integrity Signoff (무결성 서명)",
    signoff_status: "final" as const,
    chain_exit_code: chainExitCode,
    artifact_path: FREEZE_ARTIFACT,
    sections_payload_sha256: sha256Sections(s1s4),
    freeze_verify_command: "py scripts/check_logos_corpus_knowledge_freeze_manifest_v1.py",
    signed_at_utc: signedAt,
  };

  return {
    schema: "logos_inquiry_report_v1",
    version: "1.0.0",
    tier: "standard",
    output_format: "inquiry_report_v1",
    generated_at_utc: signedAt,
    research_only: payload.research_only ?? true,
    non_gating: payload.non_gating ?? true,
    send_gate: payload.send_gate ?? "HOLD",
    query: query.trim(),
    preset_id: payload.preset_id,
    query_mode: payload.query_mode ?? null,
    intake,
    sections: { ...s1s4, S5: s5 },
    governance: {
      disclaimer_ko:
        "logos inquiry Standard — Track B [HYPO] · research_only · NON_GATING. 의료·진단·투자·실거래 지시가 아닙니다.",
      forbidden_claims: ["무환각 0%", "GPT 대체", "KRV/31k 원문 전체 공개", "투자·실매매 트리거"],
      quality_basis_ko: "eval·Brier·citation lock·freeze pin — 마케팅 환각률 주장 금지.",
    },
    evidence_confidence: payload.evidence_confidence ?? null,
    azure_distill_meta: payload.azure_distill_meta ?? null,
  };
}

export { evaluateLogosTextMvpIntake as evaluateLogosInquiryIntake };
export { enrichLogosResearchQuery, hasScriptureAnchor } from "./logosInquiryQueryEnrichV1";

File: scripts/core/logos_inquiry_report_v1.py
"""Logos inquiry report v1 — Standard tier S1-S5 (C-layer bound · IP-safe)."""
from __future__ import annotations

import hashlib
import json
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from core.logos_research_text_mvp_report_v1 import (  # noqa: WPS433
    _build_school_groups,
    _extract_citation_lock_anchors,
    _split_bullets,
    _strip_hypo,
)

SCHEMA = "logos_inquiry_report_v1"
VERSION = "1.0.0"
TIER = "standard"
OUTPUT_FORMAT = "inquiry_report_v1"
FREEZE_MANIFEST_REL = "docs/final/artifacts/logos_corpus_knowledge_freeze_manifest_v1_latest.json"
FREEZE_ARTIFACT_REL = "reports/logos_inquiry_report_schema_freeze_v1_latest.json"
LEMMA_FLOOR = 290_000

ROOT = Path(__file__).resolve().parents[2]

_BOILERPLATE_RES = (
    re.compile(r"전량\s*TSK\s*교차참조", re.I),
    re.compile(r"Tier\s*C\s*로드맵", re.I),
    re.compile(r"큐레이션\s*프리셋", re.I),
    re.compile(r"^lemma bridge\s*—", re.I),
)


def _is_studio_boilerplate(text: str) -> bool:
    t = (text or "").strip()
    if not t:
        return True
    return any(rx.search(t) for rx in _BOILERPLATE_RES)


def _filter_boilerplate_bullets(items: list[str]) -> list[str]:
    return [x for x in items if not _is_studio_boilerplate(x)]


def _utc_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _read_json_rel(rel: str) -> dict[str, Any]:
    path = ROOT / rel.replace("/", "\\") if "\\" in str(ROOT) else ROOT / rel
    if not path.is_file():
        return {}
    return json.loads(path.read_text(encoding="utf-8"))


def load_freeze_lexicon_meta(*, root: Path = ROOT) -> dict[str, Any]:
    manifest_path = root / FREEZE_MANIFEST_REL
    manifest_sha256: str | None = None
    if manifest_path.is_file():
        raw = manifest_path.read_text(encoding="utf-8")
        manifest_sha256 = hashlib.sha256(raw.encode("utf-8")).hexdigest()
        manifest = json.loads(raw)
    elif root == ROOT:
        manifest = _read_json_rel(FREEZE_MANIFEST_REL)
    else:
        manifest = {}
    assets = manifest.get("assets") if isinstance(manifest.get("assets"), dict) else {}
    lemma = assets.get("lemma_verse_edges_jsonl") or {}
    sidecar = assets.get("sidecar_v2_corpus_full") or {}
    line_count = int(lemma.get("line_count") or 0)
    floor = int(lemma.get("min_line_count_floor") or LEMMA_FLOOR)
    return {
        "lemma_edge_line_count": line_count,
        "min_line_count_floor": floor,
        "floor_pass": line_count >= floor if line_count else False,
        "freeze_manifest_pointer": FREEZE_MANIFEST_REL,
        "manifest_sha256": manifest_sha256,
        "lemma_edges_sha256": str(lemma.get("sha256") or ""),
        "sidecar_corpus_sha256": str(sidecar.get("sha256") or ""),
    }


def load_sasang_regime_hint_b_track(*, root: Path = ROOT) -> dict[str, Any] | None:
    rel = "docs/final/artifacts/sasang_independent_lens_latest.json"
    path = root / rel
    if not path.is_file():
        return None
    doc = json.loads(path.read_text(encoding="utf-8"))
    stream = doc.get("sasang_stream_outputs") if isinstance(doc.get("sasang_stream_outputs"), dict) else {}
    hypothesis = str(stream.get("regime_hypothesis") or "").strip() or None
    mapping = str(stream.get("mapping_target") or "").strip() or None
    if not hypothesis and not mapping:
        return None
    return {
        "schema": "sasang_regime_hint_b_track_v1",
        "research_only": True,
        "non_gating": True,
        "regime_hypothesis": hypothesis,
        "mapping_target": mapping,
        "token": f"Sasang_Regime:{hypothesis or 'unknown'}",
        "disclaimer_ko": "[HYPO][NON_GATING] B-track 사상 렌즈 토큰만 — 임상·Track A·실매매 트리거 아님.",
    }


def _path_token_preview(payload: dict[str, Any], cap: int = 16) -> list[str]:
    path = payload.get("path") or {}
    tokens: list[str] = []
    seen: set[str] = set()

    def _push(t: str) -> None:
        v = t.strip()
        if not v or v in seen:
            return
        seen.add(v)
        tokens.append(v)

    for raw in list(path.get("node_ids") or []) + list(path.get("steps") or []):
        t = str(raw).strip()
        if re.match(r"^[A-Za-z0-9]+\.\d", t):
            _push(t)
        elif t.startswith("lemma_") or t.startswith("node:"):
            _push(t.replace("node:", ""))
        elif re.search(r"gematria", t, re.I):
            _push(t)
        if len(tokens) >= cap:
            break

    for ref in path.get("verse_refs") or []:
        if len(tokens) >= cap:
            break
        verse = str(ref).strip()
        if not verse:
            continue
        pin = hashlib.sha256(f"gematria_pin:{verse}".encode("utf-8")).hexdigest()[:12]
        _push(f"Gematria_Pin:{pin}")

    return tokens[:cap]


def _sections_payload_for_hash(sections_s1_s4: dict[str, Any]) -> str:
    canonical = json.dumps(sections_s1_s4, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    return canonical


def build_inquiry_report(
    payload: dict[str, Any],
    *,
    query: str,
    intake: dict[str, Any] | None = None,
    freeze_meta: dict[str, Any] | None = None,
    chain_exit_code: int = 0,
    artifact_path: str = FREEZE_ARTIFACT_REL,
    sasang_hint: dict[str, Any] | None = None,
) -> dict[str, Any]:
    path = payload.get("path") or {}
    answer = _strip_hypo(str(payload.get("answer") or ""))
    verse_refs = [str(v).strip() for v in (path.get("verse_refs") or []) if str(v).strip()]
    insight = payload.get("insight_card") or {}
    summary_bullets = _filter_boilerplate_bullets(_split_bullets(str(payload.get("answer") or "")))
    if isinstance(insight, dict) and insight.get("gap_ko"):
        gap = _strip_hypo(str(insight["gap_ko"]))
        if not _is_studio_boilerplate(gap):
            summary_bullets.append(gap)

    lex_meta = freeze_meta or load_freeze_lexicon_meta()
    school_groups = _build_school_groups(payload)
    regime_hint = sasang_hint if sasang_hint is not None else load_sasang_regime_hint_b_track()

    s1 = {
        "section_id": "S1_citation_lock",
        "title_ko": "Citation Lock (구절 인용)",
        "verse_refs": verse_refs[:24],
        "citation_lock_anchors": _extract_citation_lock_anchors(payload)[:24],
        "security_note_ko": "canonical verse ref·citation lock 인덱스만 — KRV/31k 원문 전문 노출 금지.",
    }
    s2 = {
        "section_id": "S2_lexicon_hash",
        "title_ko": "Lexicon Hash (원어 렉시콘)",
        "lemma_edge_line_count": lex_meta.get("lemma_edge_line_count"),
        "min_line_count_floor": lex_meta.get("min_line_count_floor", LEMMA_FLOOR),
        "floor_pass": bool(lex_meta.get("floor_pass")),
        "freeze_manifest_pointer": lex_meta.get("freeze_manifest_pointer", FREEZE_MANIFEST_REL),
        "manifest_sha256": lex_meta.get("manifest_sha256"),
        "lemma_edges_sha256": lex_meta.get("lemma_edges_sha256", ""),
        "sidecar_corpus_sha256": lex_meta.get("sidecar_corpus_sha256", ""),
        "path_token_preview": _path_token_preview(payload),
        "security_note_ko": "29만 행 원문 덤프 금지 — line_count·sha256 pin·경로 토큰 preview만.",
    }
    s3 = {
        "section_id": "S3_context_divergence",
        "title_ko": "Context Divergence (맥락/학파 분기)",
        "groups": school_groups,
        "note_ko": "[NON_GATING] conflict_context 기반 — Track A·실매매 트리거 아님.",
    }
    if regime_hint:
        s3["regime_hint_b_track"] = regime_hint
    s4 = {
        "section_id": "S4_dynamic_synthesis",
        "title_ko": "Dynamic Synthesis (고차원 통찰)",
        "body_ko": answer or "응답 본문을 생성하지 못했습니다.",
        "bullets_ko": summary_bullets[:8],
        "streaming_deferred": "active",
    }

    s1_s4 = {"S1": s1, "S2": s2, "S3": s3, "S4": s4}
    payload_canonical = _sections_payload_for_hash(s1_s4)
    payload_sha256 = hashlib.sha256(payload_canonical.encode("utf-8")).hexdigest()

    s5 = {
        "section_id": "S5_jema_integrity_signoff",
        "title_ko": "JEMA Integrity Signoff (무결성 서명)",
        "signoff_status": "final",
        "chain_exit_code": chain_exit_code,
        "artifact_path": artifact_path,
        "sections_payload_sha256": payload_sha256,
        "freeze_verify_command": "py scripts/check_logos_corpus_knowledge_freeze_manifest_v1.py",
        "signed_at_utc": _utc_now(),
    }

    intake_doc = intake or {"intake_gate": "PASS", "domain_lane": "logos", "intent_chip": "reports"}

    return {
        "schema": SCHEMA,
        "version": VERSION,
        "tier": TIER,
        "output_format": OUTPUT_FORMAT,
        "generated_at_utc": _utc_now(),
        "research_only": bool(payload.get("research_only", True)),
        "non_gating": bool(payload.get("non_gating", True)),
        "send_gate": str(payload.get("send_gate") or "HOLD"),
        "query": query.strip(),
        "preset_id": payload.get("preset_id"),
        "intake": intake_doc,
        "sections": {**s1_s4, "S5": s5},
        "governance": {
            "disclaimer_ko": (
                "logos inquiry Standard — Track B [HYPO] · research_only · NON_GATING. "
                "의료·진단·투자·실거래 지시가 아닙니다."
            ),
            "forbidden_claims": [
                "무환각 0%",
                "GPT 대체",
                "KRV/31k 원문 전체 공개",
                "투자·실매매 트리거",
            ],
            "quality_basis_ko": "eval·Brier·citation lock·freeze pin — 마케팅 환각률 주장 금지.",
        },
        "evidence_confidence": payload.get("evidence_confidence"),
    }
